### PR TITLE
Dependency handling

### LINKED ISSUE #265: Failure to build jsonff
I wanted to see how well the Fortran version was doing (I haven't really used it at all yet). I tried building my [jsonff](https://gitlab.com/everythingfunctional/jsonff) package and got the following error.

```
 + gfortran -c build/dependencies/iso_varying_string/src/ISO_VARYING_STRING.f90 -Wall -Wextra -Wimplicit-interface  -fPIC -fmax-errors=1 -g -fbounds-check -fcheck-array-temporaries -fbacktrace -Jbuild/gfortran_debug/jsonff -o build/gfortran_debug/jsonff/dependencies_iso_varying_string_src_ISO_VARYING_STRING.f90.o
 + gfortran -c build/dependencies/erloff/src/Module_m.f90 -Wall -Wextra -Wimplicit-interface  -fPIC -fmax-errors=1 -g -fbounds-check -fcheck-array-temporaries -fbacktrace -Jbuild/gfortran_debug/jsonff -o build/gfortran_debug/jsonff/dependencies_erloff_src_Module_m.f90.o
 + gfortran -c build/dependencies/erloff/src/Procedure_m.f90 -Wall -Wextra -Wimplicit-interface  -fPIC -fmax-errors=1 -g -fbounds-check -fcheck-array-temporaries -fbacktrace -Jbuild/gfortran_debug/jsonff -o build/gfortran_debug/jsonff/dependencies_erloff_src_Procedure_m.f90.o
 + gfortran -c build/dependencies/strff/src/strff.f90 -Wall -Wextra -Wimplicit-interface  -fPIC -fmax-errors=1 -g -fbounds-check -fcheck-array-temporaries -fbacktrace -Jbuild/gfortran_debug/jsonff -o build/gfortran_debug/jsonff/dependencies_strff_src_strff.f90.o
 + gfortran -c build/dependencies/erloff/src/Call_stack_entry_m.f90 -Wall -Wextra -Wimplicit-interface  -fPIC -fmax-errors=1 -g -fbounds-check -fcheck-array-temporaries -fbacktrace -Jbuild/gfortran_debug/jsonff -o build/gfortran_debug/jsonff/dependencies_erloff_src_Call_stack_entry_m.f90.o
 + gfortran -c build/dependencies/erloff/src/Call_stack_m.f90 -Wall -Wextra -Wimplicit-interface  -fPIC -fmax-errors=1 -g -fbounds-check -fcheck-array-temporaries -fbacktrace -Jbuild/gfortran_debug/jsonff -o build/gfortran_debug/jsonff/dependencies_erloff_src_Call_stack_m.f90.o
 + gfortran -c build/dependencies/erloff/src/Message_m.f90 -Wall -Wextra -Wimplicit-interface  -fPIC -fmax-errors=1 -g -fbounds-check -fcheck-array-temporaries -fbacktrace -Jbuild/gfortran_debug/jsonff -o build/gfortran_debug/jsonff/dependencies_erloff_src_Message_m.f90.o
 + gfortran -c build/dependencies/erloff/src/Error_list_m.f90 -Wall -Wextra -Wimplicit-interface  -fPIC -fmax-errors=1 -g -fbounds-check -fcheck-array-temporaries -fbacktrace -Jbuild/gfortran_debug/jsonff -o build/gfortran_debug/jsonff/dependencies_erloff_src_Error_list_m.f90.o
 + gfortran -c build/dependencies/erloff/src/Message_list_m.f90 -Wall -Wextra -Wimplicit-interface  -fPIC -fmax-errors=1 -g -fbounds-check -fcheck-array-temporaries -fbacktrace -Jbuild/gfortran_debug/jsonff -o build/gfortran_debug/jsonff/dependencies_erloff_src_Message_list_m.f90.o
 + gfortran -c build/dependencies/erloff/src/erloff.f90 -Wall -Wextra -Wimplicit-interface  -fPIC -fmax-errors=1 -g -fbounds-check -fcheck-array-temporaries -fbacktrace -Jbuild/gfortran_debug/jsonff -o build/gfortran_debug/jsonff/dependencies_erloff_src_erloff.f90.o
 + gfortran -c build/dependencies/parff/src/parff.f90 -Wall -Wextra -Wimplicit-interface  -fPIC -fmax-errors=1 -g -fbounds-check -fcheck-array-temporaries -fbacktrace -Jbuild/gfortran_debug/jsonff -o build/gfortran_debug/jsonff/dependencies_parff_src_parff.f90.o
 + gfortran -c ./src/jsonff.f90 -Wall -Wextra -Wimplicit-interface  -fPIC -fmax-errors=1 -g -fbounds-check -fcheck-array-temporaries -fbacktrace -Jbuild/gfortran_debug/jsonff -o build/gfortran_debug/jsonff/jsonff.f90.o
./src/jsonff.f90:46:17:

   46 |             join, &
      |                 1
Error: Symbol 'readfile' referenced at (1) not found in module 'strff'
compilation terminated due to -fmax-errors=1.
 Command failed
ERROR STOP 

Error termination. Backtrace:
#0  0x101fc4ebd
#1  0x101fc5b75
#2  0x101fc6f03
#3  0x101f109f8
#4  0x101efd552
#5  0x101efcc32
#6  0x101efdffd
#7  0x101ef8ddd
#8  0x101f92886
```

This error is from MacOS, but a similar one is emitted on Linux. It appears it is fetching the incorrect version of `strff`. It fetched version 1.0.0, but version 1.2.0 is specified in the `fpm.toml`. Maybe it's fetching dependencies depth first instead of breadth first?

### PR DESCRIPTION
**Pull request by awvwgk**
Take two on the dependency handling (see #251 for a recursive implementation).

- [x] create a dependency tree data type with serde functionality
- [x] allow translation from and to TOML of dependency tree
- [x] implement iterative fetching of each dependency level
- [x] use dependency tree in `build_model`
- [x] update functionality for `fpm-update` subcommand
- [x] `build_model` should use the dependency tree exclusively

Related #121
Fixes #265

*Proposed command line interface for `fpm-update`*

```
Name
    fpm-update(1) - manage project dependencies

Synopsis
    fpm update [--fetch-only] [--verbose] [--clean] [name(s)]

Description
    Manage and update project dependencies. If no dependency names are
    provided all the dependencies are updated automatically.

Options
  --clean       Ignore an existing package cache file
  --fetch-only  Only fetch dependencies, do not update existing projects
  --verbose     Show additional printout

Examples
 1. Initialize all project dependencies

    fpm update --fetch-only

 2. Update a list of dependencies

    fpm update --verbose dep1 dep2
```

*Dependency management strategy*
1. dependencies are handled iteratively
2. current project is added as first project to the dependency tree
3. all nodes in the dependency tree visited and their dependencies are added to the tree
4. repeat step 3 until all dependencies in the tree are visited
5. create a cache of the dependency tree (`build/cache.toml`)

### CODE REVIEW

**Comment by awvwgk** [comment]
I getting toward a working version here, one issue I noted is that git tags are not reliable and we have to treat them like branch names for now. One can actually force push and delete a tag with git, making it impossible to tell for *fpm* whether or not a dependency is fetched correctly without retrieving upstram. Therefore, I moved the TOML-Fortran pinning from `tag` to `rev`.

Right now the version is fully functional, while doing still a bit of redundant work for the `fpm-build` model, effectively walking the dependency tree twice, but the depth first approach of `fpm-build` will always find the already fetched dependencies from the dependency tree initialization doing no harm. I'll clean this up soon, but open the PR for feedback already.

**Review by LKedward — APPROVED** [review]
Thanks @awvwgk, this is a significant improvement. Works as expected and the implementation looks robust. :+1:

My only comment is that it would be good to include `fpm update` in the various general `fpm help` texts:

<https://github.com/fortran-lang/fpm/blob/f9e67cf176a790f768f038a45a1f30db3742c1eb/fpm/src/fpm_command_line.f90#L310>

<https://github.com/fortran-lang/fpm/blob/f9e67cf176a790f768f038a45a1f30db3742c1eb/fpm/src/fpm_command_line.f90#L326>

<https://github.com/fortran-lang/fpm/blob/f9e67cf176a790f768f038a45a1f30db3742c1eb/fpm/src/fpm_command_line.f90#L401>

**Comment by awvwgk** [comment]
Seems like I always overlook parts of the command line documentation, I have to make a checklist for creating new commands to make sure I get it added to all the right places for the next time.

**Comment by awvwgk** [comment]
I would like to move at least one of new commands (`fpm-install` / `fpm-update`) forward to focus again on the `fpm-dist` command. Let me know if there is anything blocking this PR in particular.

**Comment by LKedward** [comment]
> Let me know if there is anything blocking this PR in particular.

I'm happy for this PR to move forward; it would be good to get a second review if possible.

**Comment by LKedward** [comment]
Since this PR will close multiple issues and allow work to move forward elsewhere I think this can be merged if there are no further comments. @awvwgk are you happy to merge when you're ready?

**Comment by awvwgk** [comment]
Sounds good, I will rebase and merge this branch later today. If there are problems coming up later we can still handle those separately.

**Pull request merged**
